File: packages/portfolio-api/src/services/positions.ts
import { listAssets } from "../db/queries/assets";
import { getExchangeById } from "../db/queries/exchanges";
import { getLatestExchangeRate, getLatestPrice } from "../db/queries/prices";
import { getTransactionSummaryByAsset } from "../db/queries/transactions";
import { db } from "../db/runner";
import type { PositionRow } from "../types/api";

export function buildPositions(): { positions: PositionRow[]; lastUpdated: string | null } {
  const assets = listAssets();
  const today = new Date().toISOString().slice(0, 10);
  const latestExchangeRate = getLatestExchangeRate();
  const positions: PositionRow[] = [];

  for (const asset of assets) {
    const exchange = getExchangeById(asset.exchange_id);
    if (!exchange) continue;

    const summary = getTransactionSummaryByAsset(asset.id);
    const { unitsBought, unitsSold, totalInvested, realizedPnl } = summary;
    const unitsHeld = unitsBought - unitsSold;

    const hadActivity = unitsBought > 0 || totalInvested > 0;
    if (hadActivity && unitsHeld <= 0) continue;

    let priceResult: PositionRow["priceResult"];
    let currentValueEur = 0;

    // Cash positions are denominated in EUR in this app, so value is 1 EUR per unit.
    if (asset.type === "cash") {
      priceResult = {
        status: "ok",
        priceEur: 1,
        date: today,
        // Use latest known EUR->USD rate when available; 0 means "unknown" for UI rendering.
        exchangeRate: latestExchangeRate ?? 0,
      };
      currentValueEur = unitsHeld;
      const pnlPct = totalInvested > 0 ? ((currentValueEur - totalInvested) / totalInvested) * 100 : 0;
      positions.push({ asset, exchange, unitsHeld, totalInvestedEur: totalInvested, currentValueEur, pnlPct, realizedPnl, priceResult });
      continue;
    }

    const latestPrice = getLatestPrice(asset.id);

    if (!latestPrice) {
      priceResult = { status: "unavailable" };
    } else if (latestPrice.date === today) {
      priceResult = { status: "ok", priceEur: latestPrice.price_eur, date: latestPrice.date, exchangeRate: latestPrice.exchange_rate };
      currentValueEur = unitsHeld * latestPrice.price_eur;
    } else {
      priceResult = { status: "stale", priceEur: latestPrice.price_eur, lastKnownDate: latestPrice.date, exchangeRate: latestPrice.exchange_rate };
      currentValueEur = unitsHeld * latestPrice.price_eur;
    }

    const pnlPct = totalInvested > 0 ? ((currentValueEur - totalInvested) / totalInvested) * 100 : 0;
    positions.push({ asset, exchange, unitsHeld, totalInvestedEur: totalInvested, currentValueEur, pnlPct, realizedPnl, priceResult });
  }

  const lastUpdatedRow = db.query<{ date: string }, []>("SELECT MAX(date) as date FROM price_cache").get();
  return { positions, lastUpdated: lastUpdatedRow?.date ?? null };
}

File: fastapi/app/services/positions.py
from datetime import date as date_cls

from sqlmodel import Session

from app import crud
from app.models import (
    Asset,
    Exchange,
    GetPositionsResponse,
    PositionRow,
    PriceResultOk,
    PriceResultStale,
    PriceResultUnavailable,
)


def build_positions(session: Session) -> GetPositionsResponse:
    assets = crud.list_assets(session)
    today = date_cls.today().isoformat()
    latest_rate = crud.get_latest_exchange_rate(session)
    positions: list[PositionRow] = []

    for asset in assets:
        exchange = session.get(Exchange, asset.exchange_id)
        if not exchange:
            continue

        summary = crud.get_transaction_summary(session, asset.id)  # type: ignore[arg-type]
        units_bought = summary["units_bought"]
        units_sold = summary["units_sold"]
        total_invested = summary["total_invested"]
        units_held = units_bought - units_sold

        had_activity = units_bought > 0 or total_invested > 0
        if had_activity and units_held <= 0:
            continue

        realized_pnl = crud.compute_realized_pnl(session, asset.id)  # type: ignore[arg-type]

        if asset.type == "cash":
            price_result: PriceResultOk | PriceResultStale | PriceResultUnavailable = PriceResultOk(
                price_eur=1.0,
                date=today,
                exchange_rate=latest_rate if latest_rate is not None else 0.0,
            )
            current_value_eur = units_held
            pnl_pct = ((current_value_eur - total_invested) / total_invested * 100) if total_invested > 0 else 0.0
            positions.append(PositionRow(
                asset=asset,
                exchange=exchange,
                units_held=units_held,
                total_invested_eur=total_invested,
                current_value_eur=current_value_eur,
                pnl_pct=pnl_pct,
                realized_pnl=realized_pnl,
                price_result=price_result.model_dump(),
            ))
            continue

        latest_price = crud.get_latest_price(session, asset.id)  # type: ignore[arg-type]

        if not latest_price:
            price_result = PriceResultUnavailable()
            current_value_eur = 0.0
        elif latest_price.date == today:
            price_result = PriceResultOk(
                price_eur=latest_price.price_eur,
                date=latest_price.date,
                exchange_rate=latest_price.exchange_rate,
            )
            current_value_eur = units_held * latest_price.price_eur
        else:
            price_result = PriceResultStale(
                price_eur=latest_price.price_eur,
                last_known_date=latest_price.date,
                exchange_rate=latest_price.exchange_rate,
            )
            current_value_eur = units_held * latest_price.price_eur

        pnl_pct = ((current_value_eur - total_invested) / total_invested * 100) if total_invested > 0 else 0.0
        positions.append(PositionRow(
            asset=asset,
            exchange=exchange,
            units_held=units_held,
            total_invested_eur=total_invested,
            current_value_eur=current_value_eur,
            pnl_pct=pnl_pct,
            realized_pnl=realized_pnl,
            price_result=price_result.model_dump(),
        ))

    last_updated = crud.get_max_price_date(session)
    return GetPositionsResponse(positions=positions, last_updated=last_updated)
